### PR TITLE
dev-apis: platform: Add target for nRF5340 and nRF9160

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Add tgt_dev_apis_tfm_nrf which can be used for both nRF5340 and nRF9160.

Most of the porting is offloaded to the project that includes this.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
cc @magnev @andvib @frkv @ioannisg @anangl

**Comment by oyvindronningstad** [comment]
@kotegowder @jk-arm Please comment

**Comment by jk-arm** [comment]
we are started looking on this, will come back.

**Comment by jk-arm** [comment]
we might be able to accept the pull request, 
disclaimer: As we don't have the hardware, we are not able to verify the test suite on this target, the future bugfix/enhancement would be nordic semi responsibility. let me know your comment, otherwise, approach looks ok.
porting guide: https://github.com/ARM-software/psa-arch-tests/blob/master/api-tests/docs/porting_guide_dev_apis.md

**Comment by jk-arm** [comment]
> we might be able to accept the pull request,
> disclaimer: As we don't have the hardware, we are not able to verify the test suite on this target, the future bugfix/enhancement would be nordic semi responsibility. let me know your comment, otherwise, approach looks ok.
> porting guide: https://github.com/ARM-software/psa-arch-tests/blob/master/api-tests/docs/porting_guide_dev_apis.md

**Comment by jk-arm** [comment]
we don’t maintain any platform, the platform we maintain would be a fast model and musca platforms for our release. these changes have to maintain by you.

**Comment by oyvindronningstad** [comment]
> we don’t maintain any platform, the platform we maintain would be a fast model and musca platforms for our release. these changes have to maintain by you.

Ok, thanks for the explanation.

@magnev @andvib @ioannisg @anangl please review

**Comment by oyvindronningstad** [comment]
> 
> 
> > we don’t maintain any platform, the platform we maintain would be a fast model and musca platforms for our release. these changes have to maintain by you.
> 
> Ok, thanks for the explanation.
> 
> @magnev @andvib @ioannisg @anangl please review

^

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+/**
+ *   @brief    - Resets the system.
+ *   @param    - void
+ *   @return   - SUCCESS/FAILURE
+**/
+int pal_system_reset(void)
+{
+    /* Reset functionality is not implemented. */
+    return PAL_STATUS_UNSUPPORTED_FUNC;
```
Comment (anangl) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:158]: Why is it not supported? Most platforms call `tfm_platform_system_reset()` here. Can't we do the same?

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+/**
+    @brief    - This function parses the input string and writes bytes into UART TX FIFO
+    @param    - str      : Input String
+              - data     : Value for format specifier
+    @return   - SUCCESS/FAILURE
+**/
+
+int pal_print_ns(const char *str, int32_t data)
+{
+    logger(str, data);
```
Comment (anangl) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:61]: Why not use here `tfm_log_printf()` directly, as the psoc64 platform does? https://github.com/ARM-software/psa-arch-tests/blob/e52d4663332ca209ec3b9a2fdd1887dfaf0a791d/api-tests/platform/targets/tgt_dev_apis_tfm_psoc64/nspe/pal_driver_intf.c#L41-L43

Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: Good point, done.

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c (hunk 1 above)
Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: This was copied from somewhere. tfm_platform_system_reset() is a good idea, thanks, done.

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+void pal_timer_init_ns(uint32_t ticks);
+
+/* Start the timer. */
+void pal_timer_start_ns(void);
+
+/* Stop and resest the timer. */
+void pal_timer_stop_ns(void);
+
+/* Get the address of a free, word-aligned, 1K memory area. */
+uint32_t pal_nvmem_get_addr(void);
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: Done

**Review by anangl — APPROVED** [review]

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    @brief    - Reads from given non-volatile address.
+    @param    - base    : Base address of nvmem
+                offset  : Offset
+                buffer  : Pointer to source address
+                size    : Number of bytes
+    @return   - SUCCESS/FAILURE
+**/
+int pal_nvmem_read_ns(addr_t base, uint32_t offset, void *buffer, int size)
+{
+    base = pal_nvmem_get_addr();
```
Comment (andvib) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:110]: What's the purpose of fetching the address here? Won't that just overwrite whatever is given as an argument?

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+**/
+
+int pal_print_ns(const char *str, int32_t data)
+{
+    tfm_log_printf(str, data);
+    return PAL_STATUS_SUCCESS;
+}
+
+/**
+    @brief           - Initializes an hardware watchdog timer
```
Comment (andvib) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:62]: Small typo: Initializes _a_ hardware ...

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+extern int tfm_log_printf(const char *, ...);
+extern int32_t tfm_platform_system_reset(void);
+
+/* Initialize the timer with the given number of ticks. */
+extern void pal_timer_init_ns(uint32_t ticks);
+
+/* Start the timer. */
+extern void pal_timer_start_ns(void);
+
+/* Stop and resest the timer. */
```
Comment (andvib) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:31]: Small typo in reset

**Review by andvib — APPROVED** [review]

**Review by magnev — APPROVED** [review]

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c (hunk 6 above)
Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: Fixed

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c (hunk 5 above)
Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: Fixed

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c (hunk 4 above)
Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:114]: The base address comes from

    nvmem.0.start = 0; // Unused value;

In target.cfg. The value is not known at build time, so instead it's retrieved this way. I added a check now that ensures base is `0`.

File: api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+ *   @param    - void
+ *   @return   - SUCCESS/FAILURE
+**/
+int pal_system_reset(void)
+{
+    return tfm_platform_system_reset();
+}
+
+/* Watchdog timeout handler. */
+void TIMER1_Handler(void)
```
Comment (anangl) [inline api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c:165]: This looks strange here, as other TIMER1 related stuff is located in `plat_test.c`.

Comment (oyvindronningstad) [reply api-tests/platform/targets/tgt_dev_apis_tfm_nrf/nspe/pal_driver_intf.c]: Sure, moved it back.

**Comment by oyvindronningstad** [comment]
@jk-arm This is ready from our side now.

**Comment by jk-arm** [comment]
merging the support for nRF5340 and nRF9160, this will be maintained by nordicsemi in future

**Comment by jk-arm** [comment]
@oyvindronningstad 
done now

**Pull request merged**